### DISCUSSION TITLE
uuid attribute and element content in party-uuid?

### DISCUSSION THREAD

**Discussion by kkretsch** [usnistgov/OSCAL#1252] [Q&A]
When I look at the SSP example https://github.com/usnistgov/oscal-content/blob/main/examples/ssp/xml/ssp-example.xml I see the usage of UUID nearly everywhere as an element attribute, which I think should be the case everywhere.

But for the party-uuid element the uuid content is used inside the tags as a text node.

For parsing and validating xml I always tried to keep it in the same pattern and use only attributes, no mix.
For validating uuid references it might even be mandatory to use ID and IDREF as attribute types.

Here the two snippets out of the SSP example:

```
<party uuid="96c362ee-a012-4e07-92f3-486ab303b0e7" type="organization">
         <name>Acme Corp</name>
      </party>
```

```
<responsible-role role-id="provider">
            <party-uuid>96c362ee-a012-4e07-92f3-486ab303b0e7</party-uuid>
         </responsible-role>
```

**Comment by GaryGapinski** [comment]
I have also thought [that](https://github.com/usnistgov/OSCAL/blob/main/xml/schema/oscal_complete_schema.xsd#L4327) was an odd inconsistency. [`<member-of-organization>`](https://github.com/usnistgov/OSCAL/blob/main/xml/schema/oscal_complete_schema.xsd#L882) is similar.

**Comment by aj-stein-nist** [comment]
Not that it contradicts your position and taste re attributes, but in many cases for different OSCAL elements the `@uuid`, when it references a target to something else, is a singular non-scalar reference. In `responsible-role`, I always assumed it is different so you could have a `role` fulfilled by multiple `party-uuid`s as a list of them.

```xml
<responsible-role role-id="provider">
  <party-uuid>c86f2ccf-3391-4d4a-a6fc-785a19480399</party-uuid>
  <party-uuid>96173cd9-91e7-44d7-9b0c-4766bb08d0ac</party-uuid>
</responsible-role>
```

Now is there a reason for it to be this way and not an attribute? I don't think I have the background there (yet). It seems maybe similar for `member-of-organization`. I have nothing to back this up without researching or talking to the team, but I presumed this XML construct _could_ to more readable [JSON](https://raw.githubusercontent.com/usnistgov/oscal-content/8cf06306741f745740da65bc52d8c67adba23da8/nist.gov/SP800-53/rev5/json/NIST_SP-800-53_rev5_catalog.json) and YAML equivalents given how our Metaschema aligns the former with the latter two, but I do not think that is an actual limit to how Metaschema can define OSCAL in the different data formats, but just a hunch.

```js
{
  "catalog": {
    "uuid": "4cd70b73-a0be-4bf5-9cc6-6535dd251939",
    "metadata": {
      "roles": [
        {
          "id": "creator",
          "title": "Document creator"
        },
        {
          "id": "contact",
          "title": "Contact"
        }
      ],
      "parties": [
        {
          "uuid": "41a93829-b76b-43ec-b9e7-250553511549",
        }
       ],
      "responsible-parties": [
        {
          "role-id": "creator",
          "party-uuids": [
            "41a93829-b76b-43ec-b9e7-250553511549"
          ]
        },
        {
          "role-id": "contact",
          "party-uuids": [
            "41a93829-b76b-43ec-b9e7-250553511549"
          ]
        }
      ]
   }
}
```

**Comment by GaryGapinski** [comment]
Unsure if the choice was due to alien genetic material from JSON or YAML, but XML attribute values can be whitespace-separated NMTOKENS when more than one value is desired for an attribute (such as a fanciful `@party-uuids`).

**Comment by wendellpiez** [comment]
@GaryGapinski true, but we have to date not stipulated an NMTOKENS equivalent for OSCAL, nor are we in a rush to do so. Largely because it makes correct parsing impossible without prior knowledge from the schema or metaschema. (Personally I hate schema dependencies in parsing and I think you might too.)

In brief, the choice was due to both alien genetic material, and our need to keep the serialization both true to the model, and robust across usage scenarios (with and without schemas). This is one of many places there are uneasy compromises.

Awesome question @kkretsch.

**Comment by aj-stein-nist** [comment]
We reviewed this and believe we answered the question, and the example is correct. We will convert this to a discussion and converted it away from a bug report. We believe this is not a bug or error in the example documented cited. In the discussion, let us know if you need further clarification and discussion.
